### PR TITLE
Registration rework

### PR DESCRIPTION
**Pull request by Dalet**
Issue: #148 

Based on https://github.com/leaderboardsgg/leaderboard-backend/pull/146 (for the user view model)

- Adds validation error codes intended to be used by the front end
- Changes the HTTP status code to 422 for request validation errors

### CODE REVIEW

File: global.json
Comment (TheTedder) [inline global.json:1]: Put this in a separate PR please.

Comment (Dalet) [reply global.json:1]: https://github.com/leaderboardsgg/leaderboard-backend/pull/151
I've put it here
the CI will fail until this is merged

**Review by zysim — DISMISSED** [review]

**Review by TheTedder — CHANGES_REQUESTED** [review]
Please resolve the github actions warnings.

File: LeaderboardBackend/Migrations/20230604095631_AddUsernameAndEmailUniqueIndexes.cs
Comment (TheTedder) [inline LeaderboardBackend/Migrations/20230604095631_AddUsernameAndEmailUniqueIndexes.cs:1]: Make it so that usernames that only differ in capitalization are still treated as the same name. This can be achieved with a function index.

File: LeaderboardBackend/Models/Validation/UserPasswordRule.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+using FluentValidation;
+
+namespace LeaderboardBackend.Models.Validation;
+
+public static class UserPasswordRule
+{
+    public const string NAME = "PasswordFormat";
```
Comment (TheTedder) [inline LeaderboardBackend/Models/Validation/UserPasswordRule.cs:7]: This should probably be called PASSOWRD_FORMAT instead, or maybe RULE_NAME.

File: LeaderboardBackend/Models/Validation/UsernameRule.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+using FluentValidation;
+
+namespace LeaderboardBackend.Models.Validation;
+
+public static class UsernameRule
+{
+    public const string NAME = "UsernameFormat";
```
Comment (TheTedder) [inline LeaderboardBackend/Models/Validation/UsernameRule.cs:7]: Again, a better variable name could be used. I think USERNAME_FORMAT is good.

**Review by TheTedder — CHANGES_REQUESTED** [review]

File: LeaderboardBackend/Migrations/20230604095631_AddUsernameAndEmailUniqueIndexes.cs
Comment (zysim) [reply LeaderboardBackend/Migrations/20230604095631_AddUsernameAndEmailUniqueIndexes.cs:1]: As mentioned in Discord, code already does that, actually. I've modded the tests to verify that.

File: LeaderboardBackend/Services/Impl/UserService.cs [hunk 3]
```diff
@@ -35,9 +34,41 @@ public UserService(ApplicationContext applicationContext)
...
+        SELECT count(*) FILTER (WHERE lower(t.username) = @__ToLower_0)::int > 0 AS "usernameExists",
+            count(*) FILTER (WHERE lower(t.email) = @__ToLower_1)::int > 0 AS "emailExists"
+        FROM (
+            SELECT u.email, u.username, 1 AS "Key"
+            FROM users AS u
+        ) AS t
+        GROUP BY t."Key"
+        LIMIT 1
+        */
+        var conflicts = await _applicationContext.Users.GroupBy(x => 1)
```
[the remark this answers is not in the document]
Comment (Dalet) [reply LeaderboardBackend/Services/Impl/UserService.cs:49]: Without the `GroupBy`, the parameter for  the `Select` lambda would be a single `User` entity, so it would not be possible to call `Count()` on it.
It can be done without the `GroupBy` this way: 
```csharp
var conflicts = await _applicationContext.Users
.Select(users => new
{
    usernameExists = _applicationContext.Users.Count(x => x.Username.ToLower() == request.Username.ToLower()) > 0,
    emailExists = _applicationContext.Users.Count(x => x.Email.ToLower() == request.Email.ToLower()) > 0
}).FirstOrDefaultAsync();
```
_(note the repeated use of `_applicationContext.Users`)_
which produces the following SQL query:
```sql
SELECT (
    SELECT count(*)::int
    FROM users AS u0
    WHERE lower(u0.username) = @__ToLower_0) > 0 AS "usernameExists", (
    SELECT count(*)::int
    FROM users AS u1
    WHERE lower(u1.email) = @__ToLower_1) > 0 AS "emailExists"
FROM users AS u
LIMIT 1
```

The current SQL query for reference:
```sql
SELECT count(*) FILTER (WHERE lower(t.username) = @__ToLower_0)::int > 0 AS "usernameExists",
    count(*) FILTER (WHERE lower(t.email) = @__ToLower_1)::int > 0 AS "emailExists"
FROM (
    SELECT u.email, u.username, 1 AS "Key"
    FROM users AS u
) AS t
GROUP BY t."Key"
LIMIT 1
```

I don't know which SQL query is better (if there is even a difference), I just think the code with the `GroupBy` is more readable because it doesn't repeatedly use `_applicationContext.Users`.

[the remark this answers is not in the document]
Comment (zysim) [reply LeaderboardBackend/Services/Impl/UserService.cs:49]: I see. Thanks for the detailed info

Let's keep the `GroupBy` :+1:

File: LeaderboardBackend/Controllers/UsersController.cs [hunk 4]
```diff
@@ -51,45 +53,50 @@ public async Task<ActionResult<UserViewModel>> GetUserById(Guid id)
...
+    ///     - **Username**:
+    ///       - **UsernameTaken**: the username is already in use
+    ///     - **Email**:
+    ///       - **EmailAlreadyUsed**: the email is already in use
     /// </response>
     [AllowAnonymous]
     [HttpPost("register")]
+    [ApiConventionMethod(typeof(Conventions), nameof(Conventions.PostAnon))]
     [ProducesResponseType(StatusCodes.Status201Created)]
-    [ProducesResponseType(StatusCodes.Status400BadRequest)]
```
Comment (TheTedder) [inline LeaderboardBackend/Controllers/UsersController.cs:62]: Can't this still produce a 400 due to, say, malformed JSON?

File: LeaderboardBackend/Controllers/UsersController.cs [hunk 5]
```diff
@@ -51,45 +53,50 @@ public async Task<ActionResult<UserViewModel>> GetUserById(Guid id)
...
-    [ProducesResponseType(StatusCodes.Status409Conflict)]
+    [ProducesResponseType(StatusCodes.Status409Conflict, Type = typeof(ValidationProblemDetails))]
     public async Task<ActionResult<UserViewModel>> Register([FromBody] RegisterRequest request)
```
Comment (TheTedder) [inline LeaderboardBackend/Controllers/UsersController.cs:81]: This endpoint should be at `/account/register` per #148

File: LeaderboardBackend.Test/Features/Users/UserPasswordRuleTests.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+public class UserPasswordTests
+{
+    private readonly TestValidator _sut = new();
+
+    [TestCase("MWqVR7vHLZvizTLQX7")]
+    [TestCase("5#SP:|2^U?2PGt.fv1", Description = "contains special characters")]
+    [TestCase("cA8RlO8i", Description = "8 characters (minimum")]
+    [TestCase("9cAf9eQIt3zSNCJFfGzAX5Fu2EjCxf8r11biqEl4gKAyDcsqjyJcwkgu8hEivJoc9FQqUtGpWaNdpzVc",
+        Description = "80 characters (maximum)")]
+    public void ValidPassword(string password)
```
Comment (TheTedder) [inline LeaderboardBackend.Test/Features/Users/UserPasswordRuleTests.cs:17]: Why are these tests in a separate file from the other password tests?

File: LeaderboardBackend.Test/Features/Users/UserPasswordRuleTests.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        result.ShouldNotHaveValidationErrorFor(x => x.Password);
+    }
+
+    [TestCase("OuxNzURtWdXWd", Description = "No number")]
+    [TestCase("DZWVZVV5ED8QE", Description = "No lowercase letter")]
+    [TestCase("y267pmi50skcc", Description = "No uppercase letter")]
+    [TestCase("zmgoyGS", Description = "7 characters")]
+    [TestCase("qutOboNSzYplEKCDlCEbGPIEtMEnJImHwnluHvksTZbhuHSwFLpvUZQQxIdHctldJkdEVMRyiWcyuIeBe",
+        Description = "81 characters")]
+    public void InvalidPasswordFormat(string password)
```
Comment (TheTedder) [inline LeaderboardBackend.Test/Features/Users/UserPasswordRuleTests.cs:30]: duplicate test

File: LeaderboardBackend/Models/Requests/UserRequests.cs [hunk 8]
```diff
@@ -97,14 +90,15 @@ public record RegisterRequest
...
-    public string Password { get; set; } = null!;
+    public required string Password { get; set; }
+}
 
-    /// <summary>
-    ///     The password confirmation. This value must match `Password`.
-    /// </summary>
-    [Compare("Password", ErrorMessage = "The password confirmation must match your password.")]
-    [Required]
-    public string PasswordConfirm { get; set; } = null!;
```
Comment (TheTedder) [inline LeaderboardBackend/Models/Requests/UserRequests.cs:109]: Why has this been removed?

**Review by TheTedder — CHANGES_REQUESTED** [review]
Some changes to make and questions to answer. Still needs functional index on both email and username

EDIT: Consider using this instead https://www.postgresql.org/docs/current/citext.html

File: LeaderboardBackend/Controllers/UsersController.cs (hunk 4 above)
Comment (Dalet) [reply LeaderboardBackend/Controllers/UsersController.cs:62]: This is now included in the ApiConvention (line 78)
400 is still documented:
image

File: LeaderboardBackend/Models/Requests/UserRequests.cs (hunk 8 above)
Comment (Dalet) [reply LeaderboardBackend/Models/Requests/UserRequests.cs:109]: Ensuring that the user correctly typed their password is a client-side concern.

File: LeaderboardBackend/Models/Entities/User.cs [hunk 9]
```diff
@@ -80,3 +91,26 @@ public override int GetHashCode()
...
+        builder.HasData(
+            new User
+            {
+                Id = User.s_seedAdminId,
+                Role = UserRole.Administrator,
+                Email = "omega@star.com",
+                Password = "$2a$11$tNvA94WqpJ.O7S7D6lVMn.E/UxcFYztl3BkcnBj/hgE8PY/8nCRQe", // "3ntr0pyChaos"
+                Username = "Galactus"
+            }
+        );
```
[the remark this answers is not in the document]
Comment (zysim) [reply LeaderboardBackend/Models/Entities/User.cs:114]: It seeds the Postgres DB with an admin, so that we can run operations easier.

File: LeaderboardBackend/Models/Entities/User.cs [hunk 10]
```diff
@@ -1,14 +1,28 @@
...
-using System.ComponentModel.DataAnnotations.Schema;
 using System.Text.Json.Serialization;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
 
 namespace LeaderboardBackend.Models.Entities;
 
+public enum UserRole : byte
+{
+    None = 0, // should not be used
```
[the remark this answers is not in the document]
Comment (zysim) [reply LeaderboardBackend/Models/Entities/User.cs:10]: Yeah, I think this should be safe to remove.

File: LeaderboardBackend.Test/Features/Users/UserPasswordRuleTests.cs (hunk 7 above)
Comment (zysim) [reply LeaderboardBackend.Test/Features/Users/UserPasswordRuleTests.cs:30]: It actually isn't. This tests for invalid passwords.

File: LeaderboardBackend/Models/Entities/User.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (TheTedder) [reply LeaderboardBackend/Models/Entities/User.cs:114]: That shouldn't be done here. This literally means that everyone has admin because this code is public. This needs to be handled securely by the database admin.

File: LeaderboardBackend.Test/Features/Users/UserPasswordRuleTests.cs (hunk 6 above)
Comment (zysim) [reply LeaderboardBackend.Test/Features/Users/UserPasswordRuleTests.cs:17]: There are tests in other files? I don't see them.

Comment (zysim) [reply LeaderboardBackend.Test/Features/Users/UserPasswordRuleTests.cs:17]: If you're referring to the registration tests; I do agree there's some overlap, but functionally, those are endpoint tests, while UserPasswordRuleTests are unit tests for the password validator. I think it's fine to keep these files separate, and there's no need to consolidate them.

File: LeaderboardBackend/Controllers/AccountController.cs [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+using LeaderboardBackend.Services;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using BCryptNet = BCrypt.Net.BCrypt;
+
+namespace LeaderboardBackend.Controllers;
+
+[ApiController]
+[Route("api/[controller]")]
+public class AccountController : ControllerBase
```
[the remark this answers is not in the document]
Comment (zysim) [reply LeaderboardBackend/Controllers/AccountController.cs:14]: Oh, oops. I assumed you meant `/api/account`, implicitly.

Do we _want_ this to not be prefixed with `/api`? Sorry if I've missed/forgotten about any discussion leading to that. If so; it'll be trivial to rename the prefix

File: LeaderboardBackend/Migrations/20230624090236_Users_AddRoleAndDropAbout.cs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    public partial class UsersAddRoleAndDropAbout : Migration
+    {
+        /// <inheritdoc />
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.DropColumn(
+                name: "about",
+                table: "users");
+
+            migrationBuilder.AddColumn<byte>(
```
[the remark this answers is not in the document]
Comment (zysim) [reply LeaderboardBackend/Migrations/20230624090236_Users_AddRoleAndDropAbout.cs:19]: NPGSQL's responsible for the mapping that's been generated here. You can refer to [their docs page on it](https://www.npgsql.org/doc/types/basic.html). I'll look into this right now.

File: LeaderboardBackend/Services/Impl/UserService.cs [hunk 13]
```diff
@@ -21,23 +20,50 @@ public UserService(ApplicationContext applicationContext)
...
+            .Select(users => new
+            {
+                usernameExists = users.Count(x => x.Username == request.Username) > 0,
+                emailExists = users.Count(x => x.Email == request.Email) > 0
+            }).SingleAsync();
+
+        if (conflicts.usernameExists || conflicts.emailExists)
+        {
+            return new CreateUserConflicts(Username: conflicts.usernameExists, Email: conflicts.emailExists);
+        }
```
[the remark this answers is not in the document]
Comment (Dalet) [reply LeaderboardBackend/Services/Impl/UserService.cs:53]: The alternative is catching an exception when inserting and checking which constraint was violated, but I think only one violated constraint at a time will be reported.

File: LeaderboardBackend/Controllers/AccountController.cs [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    ///     The `LoginRequest` instance from which to perform the login.
+    /// </param>
+    /// <response code="200">
+    ///     The `User` was logged in successfully. A `LoginResponse` is returned.
+    /// </response>
+    /// <response code="400">The request was malformed.</response>
+    /// <response code="401">The password passed was incorrect.</response>
+    /// <response code="404">No `User` with the requested details could be found.</response>
+    [AllowAnonymous]
+    [HttpPost("login")]
```
[the remark this answers is not in the document]
Comment (zysim) [reply LeaderboardBackend/Controllers/AccountController.cs:92]: I'll just include #149 here. Not worth the effort to split it out.

File: LeaderboardBackend/Models/Entities/User.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (zysim) [reply LeaderboardBackend/Models/Entities/User.cs:114]: Yeah, I'm aware. Figured that since this project's an MVP it's fine; it's quicker to develop.

File: LeaderboardBackend/Controllers/AccountController.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Dalet) [reply LeaderboardBackend/Controllers/AccountController.cs:92]: true

File: LeaderboardBackend/Models/Entities/User.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Dalet) [reply LeaderboardBackend/Models/Entities/User.cs:10]: I like enums to start at 1 rather than zero because it means it was explicitly chosen, and the default value of numeric types will be invalid.
We can remove the None line and do `Registered = 1`  to achieve this though.

File: LeaderboardBackend/Services/Impl/UserService.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (TheTedder) [reply LeaderboardBackend/Services/Impl/UserService.cs:53]: That's the proper way to do it. Catch the exception and then return the corresponding error code.

File: LeaderboardBackend/Models/Entities/User.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (TheTedder) [reply LeaderboardBackend/Models/Entities/User.cs:114]: I don't think we should be committing seed data to the repo (unless it's for a local test env). This needs to be handled by whoever's in charge of the database. Everyone who needs admin access should make an account and follow the flow to be promoted to admin. Nobody should be coded to have admin.

File: LeaderboardBackend/Services/Impl/UserService.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Dalet) [reply LeaderboardBackend/Services/Impl/UserService.cs:53]: Yeah you're right, because someone could register with the same username in between the check and the insertion.

With other database providers you would have to parse the error message to get the constraint name, but with Postgres' provider it seems the violated constraint name is a property on the exception.

File: LeaderboardBackend/Controllers/AccountController.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (TheTedder) [reply LeaderboardBackend/Controllers/AccountController.cs:14]: Yes, it should be just `/account`.

File: LeaderboardBackend/Models/Entities/User.cs [hunk 15]
```diff
@@ -1,14 +1,27 @@
 using System.ComponentModel.DataAnnotations;
-using System.ComponentModel.DataAnnotations.Schema;
 using System.Text.Json.Serialization;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
 
 namespace LeaderboardBackend.Models.Entities;
 
+public enum UserRole : byte
```
Comment (TheTedder) [inline LeaderboardBackend/Models/Entities/User.cs:8]: I don't think we need to explicitly specify `byte` here. In fact, I would avoid doing so as to not mess up the DB mapping.

File: LeaderboardBackend/Migrations/20230624090236_Users_AddRoleAndDropAbout.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (TheTedder) [reply LeaderboardBackend/Migrations/20230624090236_Users_AddRoleAndDropAbout.cs:19]: It's described completely [here](https://www.npgsql.org/doc/types/enums_and_composites.html?tabs=datasource).

[the remark this answers is not in the document]
Comment (Dalet) [reply LeaderboardBackend/Migrations/20230624090236_Users_AddRoleAndDropAbout.cs]: changed

File: LeaderboardBackend/Services/Impl/UserService.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Dalet) [reply LeaderboardBackend/Services/Impl/UserService.cs]: changed

File: LeaderboardBackend/Controllers/AccountController.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Dalet) [reply LeaderboardBackend/Controllers/AccountController.cs:14]: changed

File: LeaderboardBackend/Models/Entities/User.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Dalet) [reply LeaderboardBackend/Models/Entities/User.cs:122]: Since the seeding was not added in this PR (the `HasData` part was only moved from `ApplicationContext`) imma ignore it for now

Can you open an issue for this?

[the remark this answers is not in the document]
Comment (TheTedder) [reply LeaderboardBackend/Models/Entities/User.cs:122]: Will do.

**Review by TheTedder — APPROVED** [review]
